### PR TITLE
Add -DownloadDir parameter to dotnet-install

### PR DESCRIPTION
**Pull request by jeffkl**
Our hosted build agents have multiple enlistments on them and we'd like to install the .NET Core SDK prior to building.  After the build we'd then delete the installation and re-install the version again for another build.  The current script deletes the downloaded zip file.

This change lets you specify a folder to use as a cache of downloaded installers so they can be used for installation again.

### CODE REVIEW

**Comment by dsplaisted** [comment]
Is this useful by itself?  Don't you need the option to not download if the zip file already exists, and to not delete it after it's downloaded?

**Comment by jeffkl** [comment]
> Is this useful by itself? Don't you need the option to not download if the zip file already exists, and to not delete it after it's downloaded?

That's essentially what this does.  If you specify `-DownloadDir`, the file is not downloaded again if it already exists.  So if you install SDKs over and over, they will only be downloaded once.

**Comment by dsplaisted** [comment]
Won't it keep re-downloading the file because it deletes it after each time it downloads it?

**Comment by jeffkl** [comment]
No, if you specify a `-DownloadDir` then the file is not deleted.

https://github.com/dotnet/cli/pull/10889/files#diff-7a0a71fda3f207e49f0bb7e795eebb0cR636

**Comment by dsplaisted** [comment]
Sorry, apparently I had only looked at about half of the changes.  Looks good.

**Comment by johnbeisner** [comment]
@jeffkl 
It appears that after the first download - the script is then unzipping the previously downloaded archive. Is this accurate?

**Comment by jeffkl** [comment]
> It appears that after the first download - the script is then unzipping the previously downloaded archive. Is this accurate?

Yes if the file already exists it is not downloaded so the script just unzips the previously downloaded archive.

**Comment by johnbeisner** [comment]
@jeffkl 
Unfortunately this statement would be inaccurate: "This can be useful if you will be installing the same version multiple times on the same machine."

A popular user choice with this script is the 'channel' parameter. And over the course of time, the resolved version will change - this is expected; some users want the latest version in a channel. The two parameters would not seem to mix.

I would suggest a "DownloadOnly" parameter [parameter value would be a local location] - this would put control in your hands - if your preferred version was present, no call to the install script would even be necessary.

**Comment by jeffkl** [comment]
@johnbeisner The version changes but the name of the file is unique per resolved version.  So if the version changed for a particular channel, the file wouldn't exist and would be downloaded once.

```
D:\>powershell -executionpolicy bypass -file D:\dotnet\cli\scripts\obtain\dotnet-install.ps1 -DryRun -Channel 2.0
dotnet-install: Payload URLs:
dotnet-install: Primary - https://dotnetcli.azureedge.net/dotnet/Sdk/2.1.202/dotnet-sdk-2.1.202-win-x64.zip
dotnet-install: Legacy - https://dotnetcli.azureedge.net/dotnet/Sdk/2.1.202/dotnet-dev-win-x64.2.1.202.zip

D:\>powershell -executionpolicy bypass -file D:\dotnet\cli\scripts\obtain\dotnet-install.ps1 -DryRun -Channel 2.1
dotnet-install: Payload URLs:
dotnet-install: Primary - https://dotnetcli.azureedge.net/dotnet/Sdk/2.1.503/dotnet-sdk-2.1.503-win-x64.zip
dotnet-install: Legacy - https://dotnetcli.azureedge.net/dotnet/Sdk/2.1.503/dotnet-dev-win-x64.2.1.503.zip

D:\>powershell -executionpolicy bypass -file D:\dotnet\cli\scripts\obtain\dotnet-install.ps1 -DryRun -Channel 2.2
dotnet-install: Payload URLs:
dotnet-install: Primary - https://dotnetcli.azureedge.net/dotnet/Sdk/2.2.103/dotnet-sdk-2.2.103-win-x64.zip
dotnet-install: Legacy - https://dotnetcli.azureedge.net/dotnet/Sdk/2.2.103/dotnet-dev-win-x64.2.2.103.zip

D:\>powershell -executionpolicy bypass -file D:\dotnet\cli\scripts\obtain\dotnet-install.ps1 -DryRun -Channel master
dotnet-install: Payload URLs:
dotnet-install: Primary - https://dotnetcli.azureedge.net/dotnet/Sdk/3.0.100-preview4-010554/dotnet-sdk-3.0.100-preview4-010554-win-x64.zip
dotnet-install: Legacy - https://dotnetcli.azureedge.net/dotnet/Sdk/3.0.100-preview4-010554/dotnet-dev-win-x64.3.0.100-preview4-010554.zip

D:\>powershell -executionpolicy bypass -file D:\dotnet\cli\scripts\obtain\dotnet-install.ps1 -DryRun -Channel LTS
dotnet-install: Payload URLs:
dotnet-install: Primary - https://dotnetcli.azureedge.net/dotnet/Sdk/2.1.503/dotnet-sdk-2.1.503-win-x64.zip
dotnet-install: Legacy - https://dotnetcli.azureedge.net/dotnet/Sdk/2.1.503/dotnet-dev-win-x64.2.1.503.zip
```

By specifying a download directory, its just saying to keep the downloaded file in the event it can be reused instead of downloading it every time.  So I think this feature works.  Am I missing something?

**Comment by KathleenDollard** [comment]
Seems like a good feature. 

Are we doing this on just windows? It looks like this switch is consistent with our Windows switches, but it needs lower case and a dash if it also works on Mac/Linux. 

I'm not crazy about the name. -DownloadDir sounds like an alternate location for something that would always happen, and if I understand this also initiates a check of that local cache. I don't immediately have an alternative, but even -UseDownloadDir would be a bit more active.

**Comment by jeffkl** [comment]
> I'm not crazy about the name.

I only need the functionality, so I the name of the switch is entirely up to the script maintainers.  I only named it this because there's an existing switch `-InstallDir`.  `-UseDownloadDir` seems like the name of a true/false property but the name doesn't matter to me.

@johnbeisner have you had a chance to review my response?

**Comment by johnbeisner** [comment]
@jeffkl 
I still have a few concerns.
The parameter doesn't fully explain what it is doing. Please correct me: 
1. Download the archive to a place of my choosing, unless it is already there. 
2. Install/unzip to '-InstallDir' - leave the archive, do not delete

When do the accumulated archives get reclaimed? We have initiatives in place to guard against the accumulation of SDKs on customers boxes - so in that light, this feels like a another pain point.

Whatever we do here for Windows should also be implemented for Mac/Linux.

**Comment by jeffkl** [comment]
@johnbeisner Your summation is correct of what the switch does.  The value of the switch is a directory to download the file to.  Since we can't guess where the user wants the downloads to be cached to, it makes sense for them to supply it.  Perhaps it would be more clear if you had to specify both `-DownloadDir` to just override the use of `%TEMP%` and another switch like `-KeepInstaller` to indicate you don't want to delete it.

I think if the user is opting into having the downloads saved, they would know that they are on the hook for cleaning them up.  I only need this for hosted build agents which execute numerous builds for users throughout the company.  We want to isolate individual builds by only installing their specified version but we don't want to be downloading the same version of the SDK over and over.  I don't see individual users needing this functionality.

**Comment by jeffkl** [comment]
Gentle ping

**Comment by johnbeisner** [comment]
@jeffkl 
We are getting closer; how about this:
-KeepDownloadedInstallerDirectory [value]
 Behavior:
1. If the zip/tar file is already present, do not download
2. When downloading, download to the specified directory
3. Do not delete the archive after installation
4. Works cross-platform

A bit wordy - but it sort of captures the full intent.

I must however disagree with this statement:
"I think if the user is opting into having the downloads saved, they would know that they are on the hook for cleaning them up."
As much as it logical, history has born out the opposite. At minimum a warning must be issued when reading about or using the flag.

**Comment by jeffkl** [comment]
@johnbeisner sounds good, I'll rename the switch.  Do you want me to change the message that is logged to a warning?

```powershell
if($KeepDownloadedInstallerDirectory -eq "<auto>") {
    Remove-Item $ZipPath
}
else {
    # Should this be a warning instead?
    Say "Not deleting file $ZipPath so it can be used for installation again."
}
```

**Comment by johnbeisner** [comment]
@jeffkl 
We will need this implemented in the bash script: dotnet-install.sh

If you would put the warning about responsible clean-up in the parameters documentation; I think that would be sufficient.

**Comment by johnbeisner** [comment]
@jeffkl 
Gentle ping

**Comment by jeffkl** [comment]
Sorry for the delay, other priorities have kept me from updating this PR.  I think I can get to it next week.

**Comment by jeffkl** [comment]
I'm unable to find time to complete the work for this PR.  If I get some free time I'll re-open.
